XDG Base Directory Specification adherence

elerch opened the pull request:
This change alters the default behavior for zvm to adhere to the typical XDG specification. Windows and Mac are untouched, but the groundwork has been laid to allow a more native experience on those OSs.

Having consistent behavior across all operating systems is highlighted in the project at the moment, so I understand if you may not be interested in this PR. However, I think this project does a great job, but the program behaves differently from anything else on my system under linux.

With this PR, if you set the ZVM_PATH, behavior is identical to today. If you do not, all XDG_* variables will be honored, along with their defaults.

Symlinks to zvm, zig, and zls in ~/.local/bin
Config in ~/.config/zvm
Cache in ~/.cache/zvm
Data (currently just zvm binary itself) in ~/.local/zvm/share
State (zig and zls downloads) in ~/.local/zvm/state

I have not updated the installer, and I am happy to hear any suggestions you may have. Also willing to build out the Mac/Windows conventions, but I didn't need it and I wanted to know if you were interested in this first. 😀

elerch committed:
update go.mod for newer toolchain

elerch committed:
add .envrc/.env for direnv usage

elerch committed:
unix os use XDG convention by default

tristanisham commented:
Hey Emil,

Pardon me for writing a quick response. Thanks for taking the time to work on this rework. I haven't had the time yet to review your PR so I'm mostly just responding to your post. I'll take the time to review it soon and respond with feedback, criticism, and suggestions if applicable. 


I would be interested in supporting each platform ZVM support's configuration standards, but in order to do so a few constraints would have to be met.

# 1. `upgrade` must work.

The transition to the new system would have to be flawless. User's should not be able to tell anything happened on their system. That could mean just keeping existing installs the way they are and new fresh installs could use the new path system, but backwards compatibility is really important to me. 

# 2. We'd need reliable testing for each platform

Currently ZVM really is lacking in the testing department. Part of the reason why I wrote the installer like I did was because it's easy to manage as a solo developer across the eight different operating systems ZVM supports. If we were to do this, I wouldn't feel comfortable publishing this without some kind of automated test for each platform's installer. It could be as simple as GitHub actions, and I concede that maybe not every platform will be as easy to test on, but it's important to me that the upgrade functionality does not break.

What're your thoughts on this?

elerch commented:
Yes, that makes sense. In retrospect I probably should have sent this in as a draft PR, because I wasn't sure if you were comfortable with this change overall.

The testing in particular might take some time, but I'm willing to work on that if the overall approach is ok with you. It seems as though you're ok with overall approach, so I'll work on filling in the rest.

For upgrade workflow, I think leaving existing implementations alone is better than moving a bunch of stuff around. I'll get into it and let you know how it goes.

tristanisham committed:
formatting and updated build tags

tristanisham committed:
update to be in line with master

tristanisham committed:
changed visibility to private for install functions

tristanisham committed:
something so my dumbass doesn't ship a PR

tristanisham commented:
Okay, I merged `master` with this so we're working in sync with 8.5. This worked very well on Linux by the way. I still have to test Windows and support for ZVM_PATH before I consider merging. 

Still, great work. I'm excited to work on expanding this for other platforms.

elerch committed:
Merge branch 'tristanisham:master' into xdg

elerch committed:
add gopls to .envrc

elerch committed:
add test for XDG install. Note this test actually downloads zig/zls

elerch committed:
allow processing to continue if versions.json can not be written

Realistically, it will probably error further down the line, because if it
cannot write the file, you likely have a path off. But...we have the version
data and can provide it back, just cannot persist it to disk.

elerch committed:
make sure all directories exist when installing

elerch committed:
allow skipping XDGInstall test in short mode

elerch committed:
switch to MkdirAll just in case

elerch committed:
allow action to test all operating systems

elerch committed:
support native pathing across all OSs/add tests

elerch commented:
Testing involved a refactor, but I like how it's worked out - you will want to peek at defaults_test.go. Some of the OS conventions are still a bit squishy.

I still need to work out upgrades and such as well. And documentation.

You'll want to also take a look at what I did on defaults.go, lines 31-39. I think that's what needs to be done for #77 .

elerch commented:
I should also mention I'm seeing an integration test failure for windows builds that also needs investigation

elerch committed:
add logging to diagnose actions test failure on Windows

elerch committed:
add exe extension for windows

elerch committed:
respect existing installation directories in ~/.zvm

elerch commented:
Windows integration test fixed, and added checks for existing installs.

Upgrade process, install scripts, and documentation still needed.

elerch committed:
add self field to directories to support upgrades

elerch committed:
ZVM_INSTALL support for the self field

elerch committed:
support zvm upgrade

elerch committed:
update installation scripts and add an installation go test

The installation go test will not be complete, as it does not
support the verification of changes to profile files, etc. This section
of the scripts is shell and environment dependent and I think it is
better to check manually for now

elerch committed:
update README

elerch committed:
helps if I add the env command source code

elerch committed:
skip upgrade test on mac until we can more thoroughly debug

elerch committed:
give all operating systems a shot at success

elerch committed:
skip upgrade test outside POSIX

elerch commented:
All code and documentation is ready. `install.ps1` has an issue (not reported by the tests, but you can view in the output. Also, for some reason the upgrade tests are failing on mac and Windows, so I have set those tests to skip (mac/windows only). If @tristanisham or anyone following this PR has a chance to take a look (especially on the Powershell side) I'd appreciate it. Otherwise, I do have access to those environments and can take a look sometime soon.

elerch commented:
I'll also quickly add - this PR is ready for review in my opinion. The two bugs should be minor changes.

elerch committed:
address install.ps1 issues

elerch committed:
go fmt

elerch commented:
install.ps1 has been addressed. Also ran go fmt on everything, so only upgrade on mac/windows needs attention

elerch committed:
address test failure on upgrade for Windows

The test failure was caused because the go test for upgrades does not upgrade in place,
so this should only have effected tests. Normally in place upgrades are the thing

elerch committed:
re-enable darwin as it seems to work on real hardware

elerch commented:
Windows upgrade is fixed. Mac upgrade works on actual hardware, it's just something wonky with GitHub Actions. I'll see if I can root cause this, but it should effect real users.

IMHO, this PR is ready for review merge if you're happy with it.

elerch commented:
The mac upgrade issue seems to be a flaky test. See:

https://github.com/elerch/zvm/actions/runs/13315672574/attempts/1
and 
https://github.com/elerch/zvm/actions/runs/13315672574

No code changes were made, it is purely a re-run

elerch committed:
fix flaky test on GitHub actions

This was caused by hitting rate limits, as GitHub Actions runners are pooled, they hit
limits quickly. This likely was effecting Mac and not the others due to the way Apple
requires physical hardware, so IP addresses are being reused all the time, unlike
Windows/Linux which are container based and can use unique IPs.

Three changes were made to address the issue:

1. The upgrade test was updated to fail immediately and obviously if a tag name is not
   returned
2. getLatestGitHubRelease() honors 403 and 429 failures, returning an error instead
   of silently failing
3. getLatestGitHubRelease() will use the GITHUB_TOKEN environment variable to acheive
   higher rate limits

Additionally, retry logic was put in place. This is likely unnecessary but there did not
seem to be a lot of reason to remove it. Also, some additional debug logs

elerch commented:
Flaky test is fixed. This was due to rate limits when using the GitHub REST API in getLatestGitHubRelease. While not specific to macOS, the problem surfaced there due to the way physical hardware needs to be used and shared across many jobs/repos at GitHub. More details are in the commit description

tristanisham commented:
Okay, this is awesome!!! Thank you @elerch . I'll review this as soon as possible, but it may take me a while. I'm heading out of country next week and wrapping up some work this week before I go. Thank you for putting in all this effort. I can't wait to dive into this.

isFakeAccount commented:
There is no env subcommand in zvm so the `install.sh` fails with the error.

```bash
No help topic for 'env'. help
2
/usr/bin/env
Installing zvm to 
mkdir: cannot create directory ‘’: No such file or directory
mv: cannot move '/tmp/tmp.uknwcdNhKY/zvm' to '/zvm': Permission denied
/mnt/secondary-ssd/Kira_GitProjects/GoProjects/zvm_forks/zvm/install.sh: line 92: /zvm: No such file or directory
```

tristanisham commented:
Yes! This is being implemented right now. I did decide to make some large changes to reduce scope and better fit with how I see ZVM, but you can find the edits on feat/xdg_dir_support when I push in a few.

tristanisham commented:
Continuing in #186.
